tsfc/spectral.py

```diff
@@ -68,28 +128,58 @@ def classify(argument_indices, expression):
         return COMPOUND
 
 
-def flatten(var_reps, index_cache):
-    assignments = unconcatenate([(variable, reduce(Sum, reps))
-                                 for variable, reps in var_reps],
-                                cache=index_cache)
+def delta_elimination(variable, sum_indices, args, rest):
+    """IndexSum-Delta cancellation for monomials."""
+    factors = list(args) + [variable, rest]  # construct factors
 
-    def group_key(assignment):
-        variable, expression = assignment
-        return variable.free_indices
+    def prune(factors):
+        result = remove_componenttensors(factors[:-1])
+        result.append(factors[-1])
+        return result
```

Comment:
Is it obvious why you don't want to remove component tensors from the last factor?